### PR TITLE
log-backup: don't run the cond function if the entry state is removal

### LINKED ISSUE #13044: Br log backup checkpoint lag doesn't move foward
### What version of TiKV are you using?
/ # /tikv-server -V
TiKV
Release Version:   6.2.0-alpha
Edition:           Community
Git Commit Hash:   23588581613251bdec283764a604197dd00ca377
Git Commit Branch: heads/refs/tags/v6.2.0-alpha
UTC Build Time:    2022-07-14 11:03:14

### What operating system and CPU are you using?
Cent OS

### Steps to reproduce
1. Run log backups 

### What did you expect?
1. Log backup checkpoint should move forward normally

### What did happened?
1. Log backup checkpoint doesn't move forward for 10+ hourds
M2hBVeOUv6

### PR DESCRIPTION
**Pull request by YuJuncen**
Signed-off-by: Yu Juncen <yujuncen@pingcap.com>

### What is changed and how it works?

Issue Number: Close #13044 

What's Changed:

Now, `SubscriptionManager::deregister_region_if` won't call the `cond` argument if the target has been removed by setting the state to `Removal`.

```commit-message
`SubscriptionManager::deregister_region_if` won't call the `cond` argument 
if the target has been removed by setting the state to `Removal`.
This would fix some patterns like:
let mut exists = false;
subs.deregister_region_if(42,|_, _| exists = true);
do_with(exists);
```

### Check List 

Tests 

- Integration test

### Release note 

```release-note
Fixed a bug that caused checkpoint cannot be advanced 
```

### CODE REVIEW

File: components/backup-stream/src/subscription_track.rs [hunk 1]
```diff
@@ -164,10 +164,9 @@ impl SubscriptionTracer {
         let region_id = region.get_id();
         let remove_result = self.0.get_mut(&region_id);
         match remove_result {
+            Some(o) if o.state == SubscriptionState::Removal => false,
             Some(mut o) if if_cond(o.value(), region) => {
```
Comment (overvenus) [inline components/backup-stream/src/subscription_track.rs:168]: Please do not add condition in match arm, it's very easy to make mistake, IMO. What about 
```rust
match remove_result {
  Some(mut o) => {
    if o.state == SubscriptionState::Removal {
      // ...
    }
    if if_cond(o.value(), region) {
      // ...
    }
  }
} 
```

File: components/backup-stream/src/subscription_track.rs [hunk 2]
```diff
@@ -164,10 +164,9 @@ impl SubscriptionTracer {
         let region_id = region.get_id();
         let remove_result = self.0.get_mut(&region_id);
         match remove_result {
+            Some(o) if o.state == SubscriptionState::Removal => false,
```
Comment (overvenus) [inline components/backup-stream/src/subscription_track.rs:167]: Please add some comments.

**Review by overvenus — APPROVED** [review]

File: components/backup-stream/src/subscription_manager.rs [hunk 3]
```diff
@@ -598,6 +598,15 @@ where
     }
 
     async fn get_last_checkpoint_of(&self, task: &str, region: &Region) -> Result<TimeStamp> {
+        #[cfg(feature = "failpoints")]
+        fail::fail_point!("get_last_checkpoint_of", |hint| Err(Error::Other(
```
Comment (overvenus) [inline components/backup-stream/src/subscription_manager.rs:602]: Failpoint is only enabled when `failpoint` feature is on. So no need to add `failpoints`. 
```suggestion
        fail::fail_point!("get_last_checkpoint_of", |hint| Err(Error::Other(
```

Comment (YuJuncen) [reply components/backup-stream/src/subscription_manager.rs:602]: `fail` in the crate is an optional dependency which would only be required when `failpoints` feature enabled. I am afraid if we remove the config it cannot compile...

Comment (YuJuncen) [reply components/backup-stream/src/subscription_manager.rs:602]: See:
https://github.com/tikv/tikv/blob/21e3bd64e8de517e09836cda4991b446914f3f75/components/backup-stream/Cargo.toml#L11

**Pull request merged**